src/main/java/org/jenkinsci/bytecode/helper/TypeHierarchyReader.java

```diff
@@ -0,0 +1,432 @@
+ * methods from {@link Class}. Users are expected to decorate or subclass with a
+ * caching strategy which best suits your environment or application.
+ * 
+ * @author Graham Allan
+ * 
+ */
+class TypeHierarchyReader {
+
+    /**
+     * Returns true if the given {@link Type} represents an interface.
+     * 
+     * @see Class#isInterface()
+     */
+    public boolean isInterface(final Type t) {
+        return hierarchyOf(t).isInterface();
+    }
+
+    /**
+     * <p> Returns a {@link Type} representing the superclass of the given
+     * {@link Type} <code>t</code>. </p> <p> Semantically equivalent to
+     * {@link Class#getSuperclass()}. That is, with {@link Class}es X and Y
+     * where <code>X.getSuperclass()</code> returns <code>Y</code>, and the
+     * {@link Type}s <code>tX</code> and <code>tY</code> representing
+     * <code>X</code> and <code>Y</code> respectively,
+     * <code>this.getSuperClass(tX)</code> returns <code>tY</code>. It matches
+     * {@link Class#getSuperclass()}'s behaviour for primitive, array and void
+     * types, where the given {@link Type}'s have the equivalent
+     * {@link Type#getSort()} values. </p>
+     * 
+     * @see Class#getSuperclass()
+     * @see Type#getSort()
+     */
+    public Type getSuperClass(final Type t) {
+        return hierarchyOf(t).getSuperType();
+    }
+
+    /**
+     * <p> Determines if the {@link Type} <code>to</code> is the same class, a
+     * superclass, or superinterface of the given {@link Type}
+     * <code>from</code>. Semantically equivalent to
+     * {@link Class#isAssignableFrom(Class)}, where <code>to</code> represents
+     * <code>this</code> {@link Class} object. That is, with any given
+     * {@link Class}es X and Y, and {@link Type}s tX and tY, the following
+     * statements will always be true: <br> <code>X.isAssignableFrom(Y) ==
+     * this.isAssignableFrom(tX, tY)</code> </p>
+     * 
+     * Note that, in accordance with {@link Class#isAssignableFrom(Class)}, this
+     * method does not return true for {@link Type}'s representing primitives
+     * that can be converted (i.e. widened or narrowed) to represent the same
+     * type.
+     * 
+     * @return true if instances of from can be legally assigned to instances of
+     *         to, else false
+     */
+    public boolean isAssignableFrom(Type to, Type from) {
+        return hierarchyOf(to).isAssignableFrom(hierarchyOf(from), this);
+    }
+
+    public String getCommonSuperClass(final String type1, final String type2) {
+        Type c = Type.getObjectType(type1);
+        Type d = Type.getObjectType(type2);
+
+        if (isAssignableFrom(c, d)) {
+            return type1;
+        }
+        if (isAssignableFrom(d, c)) {
+            return type2;
+        }
+        if (isInterface(c) || isInterface(d)) {
+            return "java/lang/Object";
+        } else {
+            do {
+                c = getSuperClass(c);
```

Comment:
you are always getting the superclass.  So unless the superclass is the current class then you will be fine - there are no infinite loops.

If the current class is the same as the super class you have bigger issues than a recursive loop.